### PR TITLE
feat: Change Capture mode to allow polling for any envelope

### PR DESCRIPTION
**Pull request by timfish**
This PR:
- Allows the cli to start in capture mode
- Adds a `/api/relay/last-event/` endpoint that returns the last envelope or error captured via Capture mode without the need to supply an event id.

Then in any language you just have to create a wrapper that starts the Relay process (`relay run --mode capture`) and then polls the new endpoint. 

For JavaScript testing I would create a module that downloads and caches Relay from GitHub releases and wraps the process.

### CODE REVIEW

**Comment by timfish** [comment]
I've got a demo of this working in the JavaScript repo. 

The only major issue there is that session envelopes are not currently collected in capture mode.

**Review by untitaker — APPROVED** [review]

**Comment by HazAT** [comment]
cc @tonyo @RaduW this could be the start of the SDK Payload tester

**Comment by timfish** [comment]
It looks like this is why capture mode does not capture sessions:
https://github.com/getsentry/relay/blob/b036848850f80ea7b7054b720a9569768a31d781/relay-server/src/actors/envelopes.rs#L2138-L2143

**Comment by timfish** [comment]
For this PR to fully fulfil its purpose of capturing all envelopes for testing SDKs, it also needs to capture sessions. I have a reasonable idea of how to do this but I don't know where Sentry currently uses existing capture mode functionality and what I might break!

File: relay-server/src/endpoints/events.rs [hunk 1]
```diff
@@ -28,9 +29,73 @@ fn get_captured_event(event_id: Path<EventId>) -> ResponseFuture<HttpResponse, M
...
+        }
+    })
+}
+
+fn get_last_event(_: ()) -> ResponseFuture<HttpResponse, MailboxError> {
+    let future = EnvelopeManager::from_registry()
+        .send(GetCapturedEnvelope { event_id: None })
+        .map(|result| match result {
+            Some(result) => {
+                let envelope = envelope_as_json(result);
```
Comment (jan-auer) [inline relay-server/src/endpoints/events.rs]: Why not just return the envelope as envelope? :) 
They are already a serializable and parsable format able to contain all sorts of data, including minidumps and soon metrics.

File: relay-server/src/endpoints/events.rs [hunk 2]
```diff
@@ -28,9 +29,73 @@ fn get_captured_event(event_id: Path<EventId>) -> ResponseFuture<HttpResponse, M
...
+
+    Box::new(future)
+}
+
 pub fn configure_app(app: ServiceApp) -> ServiceApp {
     app.resource("/api/relay/events/{event_id}/", |r| {
         r.name("internal-events");
         r.method(Method::GET).with(get_captured_event);
     })
+    .resource("/api/relay/last-event/", |r| {
```
Comment (jan-auer) [inline relay-server/src/endpoints/events.rs]: I'd call this something like `last-request` so we can also return more information about the latest request. Specifically, if we drop an envelope, we'll likely want to retain the outcome and return it here.

**Review by jan-auer — COMMENTED** [review]
The only unfortunate thing about `/last-event/` is that it can't be used in concurrent testing effectively. Given that Relay starts up rather quickly in capture mode, that shouldn't be too big of a problem. 

This is also the reason why we skipped session and metrics envelopes. We simply don't have an event ID to associate to. We could assign artificial IDs, but that would change the behavior of the public interface.

**Comment by timfish** [comment]
I've had a go at improving this. A few weeks ago I had a brief chat with @jan-auer who (if I remember correctly) suggested that the current capture mode is not used anywhere so I replaced it with something that should also work for non event envelopes as it doesn't rely on event ids.

The single endpoint (`/api/relay/get-envelope/`) returns raw envelopes in the order they were received and will alternatively return only the JSON elements if you request json via the `accept` header.

**Comment by iker-barriocanal** [comment]
I'm closing this pull request due to staleness.

**Pull request closed without merging**